### DISCUSSION TITLE
Can't get a .png file to show up in splashscreen as a "AndroidWindowSplashScreenAnimatedIcon" on cordova-android@12. It only shows the default Cordova splashscreen.

### DISCUSSION THREAD

**Discussion by dylanfpaul** [apache/cordova#430] [Q&A] [answered]
As the title suggests, I am trying to use a .png file for a splashscreen icon as I do not have access to vectors. I have tried using a .png with the dimensions of 960 x 960 px among other sizes to no avail. No matter what I try and implement, my app will always show the default cordova splashscreen. I have tested this on a physical Samsung S9 (Android 10) as well as many emulators ranging from Android 11 to Android 13. All cases have only shown the default cordova splashscreen.

For some context I am on `cordova-android@12`, `cordova@11`, and `Node 14.19.0`.

Here is my `config.xml` (Note: My concern here is the android side of things):

```
<?xml version='1.0' encoding='utf-8'?>
<widget android-versionCode="184" id="com.testapp.mobile" ios-CFBundleVersion="184" version="2.15.01" xmlns="http://www.w3.org/ns/widgets" xmlns:cdv="http://cordova.apache.org/ns/1.0">
    <name>My App</name>
    <description>
		test description
	</description>
    <author email="infotestxyz123@gmail.com" href="http://google.com">
		Test Author
	</author>
    <content src="index.html" />
    <ignore entries="node_modules" />
    <feature name="CDVWKWebViewEngine">
        <param name="ios-package" value="CDVWKWebViewEngine" />
    </feature>
    <preference name="CordovaWebViewEngine" value="CDVWKWebViewEngine" />
    <preference name="WKWebViewOnly" value="true" />
    <preference name="DisallowOverscroll" value="true" />
    <preference name="StatusBarOverlaysWebView" value="false" />
    <preference name="StatusBarBackgroundColor" value="#006384" />
    <preference name="StatusBarStyle" value="lightcontent" />
    <preference name="AndroidPersistentFileLocation" value="Internal" />
    <preference name="AllowInlineMediaPlayback" value="true" />
    <preference name="MediaPlaybackRequiresUserAction" value="false" />
    <access origin="*" />
    <allow-intent href="http://*/*" />
    <allow-intent href="https://*/*" />
    <allow-intent href="tel:*" />
    <allow-intent href="sms:*" />
    <allow-intent href="mailto:*" />
    <allow-intent href="geo:*" />
    <allow-navigation href="https://*.youtube-nocookie.com" />
    <allow-navigation href="https://*.youtube.com" />
    <allow-navigation href="https://*.ytimg.com" />
    <allow-navigation href="https://*.gstatic.com" />
    <allow-navigation href="https://*.googlevideo.com" />
    <allow-navigation href="https://*.google.com" />
    <allow-navigation href="https://*.vimeo.com" />
    <allow-navigation href="https://*.vimeocdn.com" />
    <hook src="hooks/before_platform_add/show_entries.js" type="before_platform_add" />
    <hook src="hooks/before_build/show_entries.js" type="before_build" />
    <platform name="android">
        <hook src="scripts/android/androidBeforePrepare.js" type="before_prepare" />
        <hook src="scripts/android/androidBeforeCompile.js" type="before_compile" />
        <hook src="scripts/hideEntriesBeforeCompile.js" type="before_compile" />
        <hook src="scripts/android/androidAfterPlatformAdd.js" type="after_platform_add" />
        <hook src="scripts/android/removeAndroidPermissions.js" type="after_prepare" />
        <allow-intent href="market:*" />
        <preference name="android-targetSdkVersion" value="33" />
        <preference name="android-minSdkVersion" value="24" />
        <preference name="AndroidInsecureFileModeEnabled" value="true" />
        <preference name="AndroidXEnabled" value="true" />
        <icon density="ldpi" src="res/android/icon/ldpi.png" />
        <icon density="mdpi" src="res/android/icon/mdpi.png" />
        <icon density="hdpi" src="res/android/icon/hdpi.png" />
        <icon density="xhdpi" src="res/android/icon/xhdpi.png" />
        <icon density="xxhdpi" src="res/android/icon/xxhdpi.png" />
        <icon density="xxxhdpi" src="res/android/icon/xxxhdpi.png" />
        <preference name="AndroidWindowSplashScreenAnimatedIcon" src="res/android/screen/splashicon.png" />
        <preference name="AndroidWindowSplashScreenBackground" value="#ffffff" />
        <preference name="SplashScreenDelay" value="1000" />
        <preference name="AutoHideSplashScreen" value="true" />
        <preference name="FadeSplashScreen" value="true"/>
        <preference name="FadeSplashScreenDuration" value="2000"/>
        <preference name="GradlePluginGoogleServicesEnabled" value="true" />
        <!-- <preference name="GradlePluginGoogleServicesVersion" value="4.3.15" /> -->
    </platform>
    <platform name="ios">
        <config-file target="*-Info.plist" parent="NSCameraUsageDescription">
            <string>Uses camera to upload images for profile pictures or for use in a post.</string>
        </config-file>
        <config-file target="*-Info.plist" parent="NSPhotoLibraryUsageDescription">
            <string>Uses photo library to upload images for profile pictures or for use in a post.</string>
        </config-file>
        <config-file target="*-Info.plist" parent="NSMicrophoneUsageDescription">
            <string>Uses camera to record video to attach to a form.</string>
        </config-file>
        <config-file target="*-Info.plist" parent="ITSAppUsesNonExemptEncryption">
            <false />
        </config-file>
        <hook src="scripts/ios/iosBeforePrepare.js" type="before_prepare" />
        <hook src="scripts/ios/iosBeforeCompile.js" type="before_compile" />
        <hook src="scripts/hideEntriesBeforeCompile.js" type="before_compile" />
        <hook src="scripts/ios/iosBeforeBuild.js" type="before_build" />
        <allow-intent href="itms:*" />
        <allow-intent href="itms-apps:*" />
        <preference name="Orientation" value="all" />
        <preference name="deployment-target" value="16.1" />
        <preference name="InAppBrowserOverrideUserAgent" value="Mozilla/5.0 (iPad; CPU OS 14_4_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1" />
        <icon height="20" src="res/ios/icon/Icon-20.png" width="20" />
        <icon height="40" src="res/ios/icon/Icon-20@2x.png" width="40" />
        <icon height="60" src="res/ios/icon/Icon-20@3x.png" width="60" />
        <icon height="29" src="res/ios/icon/Icon-29.png" width="29" />
        <icon height="58" src="res/ios/icon/Icon-29@2x.png" width="58" />
        <icon height="87" src="res/ios/icon/Icon-29@3x.png" width="87" />
        <icon height="40" src="res/ios/icon/Icon-40.png" width="40" />
        <icon height="80" src="res/ios/icon/Icon-40@2x.png" width="80" />
        <icon height="120" src="res/ios/icon/Icon-40@3x.png" width="120" />
        <icon height="50" src="res/ios/icon/Icon-50.png" width="50" />
        <icon height="100" src="res/ios/icon/Icon-50@2x.png" width="100" />
        <icon height="57" src="res/ios/icon/Icon-57.png" width="57" />
        <icon height="114" src="res/ios/icon/Icon-57@2x.png" width="114" />
        <icon height="120" src="res/ios/icon/Icon-60@2x.png" width="120" />
        <icon height="180" src="res/ios/icon/Icon-60@3x.png" width="180" />
        <icon height="72" src="res/ios/icon/Icon-72.png" width="72" />
        <icon height="144" src="res/ios/icon/Icon-72@2x.png" width="144" />
        <icon height="76" src="res/ios/icon/Icon-76.png" width="76" />
        <icon height="152" src="res/ios/icon/Icon-76@2x.png" width="152" />
        <icon height="167" src="res/ios/icon/Icon-83@2x.png" width="167" />
        <icon height="120" src="res/ios/icon/Icon-120.png" width="120" />
        <icon height="29" src="res/ios/icon/Icon-Small.png" width="29" />
        <icon height="58" src="res/ios/icon/Icon-Small@2x.png" width="58" />
        <icon height="87" src="res/ios/icon/Icon-Small@3x.png" width="87" />
        <icon height="1024" src="res/ios/icon/iTunesArtwork@2x.png" width="1024" />
        <!-- Support the new launch storyboard images. -->
        <splash src="res/ios/screen/Default@2x~iphone~anyany.png" />
        <splash src="res/ios/screen/Default@2x~iphone~comany.png" />
        <splash src="res/ios/screen/Default@2x~iphone~comcom.png" />
        <splash src="res/ios/screen/Default@3x~iphone~anyany.png" />
        <splash src="res/ios/screen/Default@3x~iphone~anycom.png" />
        <splash src="res/ios/screen/Default@3x~iphone~comany.png" />
        <splash src="res/ios/screen/Default@2x~ipad~anyany.png" />
        <splash src="res/ios/screen/Default@2x~ipad~comany.png" />
    </platform>
</widget>
```

Now I do also have these warnings in my build output, not sure if this is affecting the issue though:
![image](https://github.com/apache/cordova/assets/18554911/05b3ea40-e44f-453a-8120-c39929e1e8d9)

Here is the default cordova splashscreen showing up on an emulated Pixel 6 Pro API 33 (even after killing the application and then running it again from the launcher):
![image](https://github.com/apache/cordova/assets/18554911/c6493481-1359-4bd0-9a59-835893e671db)

I notice that the drawable assets of the default cordova splashscreen are always making it into my final application:
![image](https://github.com/apache/cordova/assets/18554911/8ec62316-4d7b-4ae9-8ab3-a126047c9318)

I have tested this on a physical Samsung S9 (android 10) as well as some other emulators to no avail. It is always just the default cordova splashscreen.

**Comment by breautek** [comment]
Can you try in a sample project without all the extra bits?

e.g. with no hooks, or plugins, just simply a pure cordova app with your splashscreen resource. If it doesn't work, then you can zip or publish that test project to a public repo where I can take a look.

If splashscreens does work, then you can incrementally start adding back your bits to see when it breaks. The hooks would be the first thing I'd look at since they have the capability of messing with the native project perhaps in an incompatible way.
Then the second likely culprit is an incompatible plugin.

I notice you have a lot of legacy stuff in your config.xml that either suggests you're using out-dated dependencies, or you have dependencies updated but simply have never updated your config. Stuff like `WKWebViewOnly` preference is a no-op (always true on cordova-ios 6.x and later) as an example. None of this really concerns the splashscreen issue, **BUT** if you have `cordova-plugin-splashscreen` installed, then you can probably uninstall it. If you use the `browser` platform, then make sure it's completely up to date, which wouldn't have an iOS or android implementation.

Splashscreen has been moved to the core platform, so if you're using the splashscreen plugin with an android implementation, it could be overwriting or conflicting with the core version of the plugin.

**Reply by dylanfpaul** [reply]
Okay, I actually have the splash screen plugin uninstalled, and all references removed from fetch.json, my plugins folder, package/package-lock etc. However I do remember that I had to remove the plugin manually because when I was uninstalling it, it said something like another deprecated plugin was not found. For instance, I think it said when trying to uninstall cordova-plugin-splashscreen "failed to uninstall cordova-plugin-splashscreen, cordova-plugin-androidx could not be found".

This being said, I will try and run the build without hooks and other things and see if I can get the splashscreen to work. I am also wondering if I should try using cordova CLI 12 instead of cordova CLI 11. I shall post here what I find out.

**Reply by breautek** [reply]
if you had cordova-plugin-androidx installed, it can be uninstalled too.

cordova only supports androidx and the old flag to enable androidx is effectively forcefully on.

> I am also wondering if I should try using cordova CLI 12 instead of cordova CLI 11

Unless if you have a really good reason not to, it's generally better to use the latest version available. Good practice to read the release notes to know what is changing in a breaking manner.

For the CLI the breaking changes are usually just dropping support for out-dated/going out of support nodejs versions. So as long as you're using the current LTS version at the very least, you'll be good usually.

**Reply by dylanfpaul** [reply]
I am about to do all of this testing today but one question I have is, should it even be possible to see the default cordova splashscreen with the deprecated cordova-plugin-splashscreen uninstalled?

**Reply by erisu** [reply]
Yes you will still see the Cordova's default splash screen assets if you do not override it with the `AndroidWindowSplashScreenAnimatedIcon` preference flag.

Eventhough the cordova-plugin-splashscreen is deprecated and should not be installed, the Splash Screen functionality still exists in the core of Cordova-Android.

**Comment by dylanfpaul** [comment]
@breautek 

[cordova_test_project.zip](https://github.com/apache/cordova/files/12527461/cordova_test_project.zip)

Alright, here is a test project with no plugins, just the android platform, and my icons / splashscreen icon that I want to use. It is still only showing the default cordova splashscreen when emulating in Android Studio, even when I kill the app after first launch and then launch it again from the launcher.

I am using **Node v14.19.0** and **cordova@11.0.0**

My global packages:

├── **@mapbox/node-pre-gyp@1.0.8**
├── **cordova@11.0.0**
├── **corepack@0.10.0**
├── **node-gyp@9.0.0**
└── **npm@6.14.16**

Here are my environment variables:

```
export NVM_DIR="$HOME/.nvm"
[ -s "/usr/local/opt/nvm/nvm.sh" ] && . "/usr/local/opt/nvm/nvm.sh"  # This loads nvm
[ -s "/usr/local/opt/nvm/etc/bash_completion.d/nvm" ] && . "/usr/local/opt/nvm/etc/bash_completion.d/nvm"  # This loads nvm bash_completion
export ANDROID_HOME=/Users/dylanpaul/Library/Android/sdk/
export JAVA_HOME=$(/usr/libexec/java_home -v 11.0.14)
export PATH="/Users/dylanpaul/bin/Sencha/Cmd:$PATH"
export PATH=$PATH:$ANDROID_HOME/platform-tools/
export PATH=$PATH:$ANDROID_HOME/tools/bin
export PATH=$PATH:$ANDROID_HOME/tools
export PATH=$PATH:$ANDROID_HOME/cmdline-tools/latest/bin/
export PATH=$PATH:$ANDROID_HOME/emulator/
export PATH=$PATH:/Users/dylanpaul/Downloads/gradle-8.3/bin
```

Here is my Android Studio info:
![image](https://github.com/apache/cordova/assets/18554911/71b5aaf1-254b-4b22-b9d2-352e5c9fa9ec)

Here is my build output:
![image](https://github.com/apache/cordova/assets/18554911/ea670caa-da92-41d3-96e9-52f37235880b)

**Comment by breautek** [comment, marked as answer]
I missed this the first time around but in your config you have

```
<preference name="AndroidWindowSplashScreenAnimatedIcon" src="res/android/screen/splashicon.png" />
```

the preference uses `value`, not `src` so it should be:

```
<preference name="AndroidWindowSplashScreenAnimatedIcon" value="res/android/screen/splashicon.png" />
```

With that being said, your splashscreen icon still doesn't look quite right. It's because you aren't obeying the mask. The canvas size is correct at 960x960, but only 640px radius from the center is usable.

I took your icon and placed a circle mask as an example:

![maskSample](https://github.com/apache/cordova/assets/10145915/fb14c8fc-b216-4cc2-a1d2-95d05f651c16)

Your icon should not exceed the circle area, so as you can see your icon content is too large.

I used inkscape to resize the PNG down (keeping the aspect ratio) to fill the entire mask:

![maybeThisWillDo](https://github.com/apache/cordova/assets/11200662/678b2c21-cef8-4940-a5f4-1ff9715472ec)

But it might be better quality wise if you do it with the original project file for your icon.

It should look something like this when you're done:

[Screencast from 2023-09-05 07:19:07 PM.webm](https://github.com/apache/cordova/assets/11200662/88e2b42f-f911-443b-8f0b-c61dfb115ec6)

**Reply by dylanfpaul** [reply]
Thank you, that preference change solved my problem. Now I just need to get the correct scaling for my icon.  I will follow your guidance about the mask.
